src/rust/engine/src/rule_graph.rs

```diff
@@ -251,115 +322,472 @@ impl<'t> GraphMaker<'t> {
+    let mut diagnostics = Vec::new();
+    for available_params in params_powerset {
+      let available_params = available_params.into_iter().collect();
+      // If a subset of these parameters is already satisfied, skip. This has the effect of
+      // selecting the smallest sets of parameters that will satisfy a rule.
+      // NB: This scan over satisfied sets is linear, but should have a small N.
+      if combinations
+        .keys()
+        .any(|satisfied_entry| satisfied_entry.params().is_subset(&available_params))
+      {
+        continue;
+      }
+
+      match Self::choose_dependencies(&available_params, &deps) {
+        Ok(Some(inputs)) => {
+          let mut rule_edges = RuleEdges::default();
+          for (key, input) in inputs {
+            rule_edges.add_edge(key.clone(), input.clone());
+          }
+          combinations.insert(entry.simplified(available_params), rule_edges);
+        }
+        Ok(None) => {}
+        Err(diagnostic) => diagnostics.push(diagnostic),
+      }
+    }
+
+    // If none of the combinations was satisfiable, return the generated diagnostics.
+    if !combinations.is_empty() {
+      Ok(combinations)
+    } else {
+      Err(diagnostics)
+    }
+  }
+
+  ///
+  /// Given a set of available Params, choose one combination of satisfiable Entry dependencies if
+  /// it exists (it may not, because we're searching for sets of legal parameters in the powerset
+  /// of all used params).
+  ///
+  /// If an ambiguity is detected in rule dependencies (ie, if multiple rules are satisfiable for
+  /// a single dependency key), fail with a Diagnostic.
+  ///
+  fn choose_dependencies<'a>(
+    available_params: &ParamTypes,
+    deps: &'a Vec<(SelectKey, Vec<Entry>)>,
+  ) -> Result<Option<Vec<(&'a SelectKey, &'a Entry)>>, Diagnostic> {
+    let mut combination = Vec::new();
+    for (key, input_entries) in deps {
+      let satisfiable_entries = input_entries
+        .iter()
+        .filter(|input_entry| {
+          Self::used_params(key, input_entry)
+            .iter()
+            .all(|p| available_params.contains(p))
+        })
+        .collect::<Vec<_>>();
+
+      let chosen_entries = Self::choose_dependency(satisfiable_entries);
+      match chosen_entries.len() {
+        0 => {
+          return Ok(None);
+        }
+        1 => {
+          combination.push((key, *chosen_entries.last().unwrap()));
+        }
+        _ => {
+          return Err(Diagnostic::ambiguous(available_params, key, chosen_entries));
+        }
+      }
+    }
+
+    Ok(Some(combination))
+  }
+
+  fn choose_dependency<'a>(satisfiable_entries: Vec<&'a Entry>) -> Vec<&'a Entry> {
+    if satisfiable_entries.is_empty() {
+      // No source of this dependency was satisfiable with these Params.
+      return vec![];
     }
-    result
+
+    // Prefer a Singleton, then a Param, then the non-ambiguous rule with the smallest set of
+    // input Params.
+    // TODO: We should likely prefer Rules to Params.
```

Comment:
Why should we prefer `Rules` to `Params`? In fact; why are these directly comparable at all? They feel like they're separate; what's an `@rule` signature with an example of a `Params` and `Rules` where we need to resolve between the two?

Reply:
Good question. This needs unification with `rhs`, but: the non-intuitive thing I discovered while implementing this is that in order to keep `Node` identities small (ie, containing the fewest Params possible), we want to prefer Entries that need fewer parameters. This biases toward avoiding propagating information from callers to callees unless they really need it to be executed (and avoids the case where absolutely anything used anywhere in the graph is potentially a parameter used at the root of the graph: that will continue to be bounded by `RootRule`).

In short: we would generally prefer a `@rule` to a Param, because depending on a rule does not affect our identity, while using a `Param` does (because it comes from the caller).

Reply:
Maybe add this comment as a code comment?